src/lfortran/semantics/ast_symboltable_visitor.cpp

```diff
@@ -838,22 +838,26 @@ class SymbolTableVisitor : public CommonVisitor<SymbolTableVisitor> {
 
         ASRUtils::SymbolDuplicator symbol_duplicator(al);
         ASRUtils::ExprStmtWithScopeDuplicator exprstmt_duplicator(al, current_scope);
-
-        for (auto &item : proc_interface->m_symtab->get_scope()) {
-            if (ASR::is_a<ASR::Variable_t>(*item.second)) {
-                ASR::Variable_t* proc_interface_vari = ASR::down_cast<ASR::Variable_t>(item.second);
-                ASR::symbol_t* new_sym = symbol_duplicator.duplicate_Variable(proc_interface_vari, current_scope);
-                current_scope->add_symbol(std::string(ASRUtils::symbol_name(new_sym)), new_sym);
-            }
-        }
-
+        symbol_duplicator.duplicate_SymbolTable(proc_interface->m_symtab, current_scope);
         Vec<ASR::expr_t*> new_func_args;
         new_func_args.reserve(al, proc_interface->n_args);
         for (size_t i=0;i<proc_interface->n_args;i++) {
             new_func_args.push_back(al, exprstmt_duplicator.duplicate_expr(proc_interface->m_args[i]));
         }
         ASR::expr_t* new_func_return_var = exprstmt_duplicator.duplicate_expr(proc_interface->m_return_var);
 
+        for (size_t i=0; i<x.n_decl; i++) {
+            is_Function = true;
+            if (!AST::is_a<AST::Require_t>(*x.m_decl[i])) {
+                try {
+                    visit_unit_decl2(*x.m_decl[i]);
+                } catch (SemanticAbort &e) {
+                    if ( !compiler_options.continue_compilation ) throw e;
+                }
+            }
+            is_Function = false;
+        }
+
```

Comment:
you may wish to preserve last state of `is_Function`

```suggestion
        bool is_Function_copy = is_Function;
        for (size_t i=0; i<x.n_decl; i++) {
            is_Function = true;
            if (!AST::is_a<AST::Require_t>(*x.m_decl[i])) {
                try {
                    visit_unit_decl2(*x.m_decl[i]);
                } catch (SemanticAbort &e) {
                    if ( !compiler_options.continue_compilation ) throw e;
                }
            }
            is_Function = false;
        }
        is_Function= is_Function_copy;
```

Reply:
I don't think `is_Function` is required anywhere else apart from DeclarationUtil. So, I don't think that we need to preserve previous state. For `visit_Subroutine` and `visit_Function` also, we do the similar thing instead of creating a copy.

Reply:
okay, still please bring the `is_Function = true` outside loop.